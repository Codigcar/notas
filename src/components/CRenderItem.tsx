import React from 'react';
import {Text, View, StyleSheet} from 'react-native';
import Proptypes from 'prop-types';

import {IPublication} from '../interfaces/publish.interface';

export const CRenderItem = ({title, body, userId}: IPublication) => {
  return (
    <View accessibilityRole='menuitem' style={styles.container}>
      <Text style={styles.title}>{title}</Text>
      <Text>{body}</Text>
      <Text>User: {userId}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 10,
    borderBottomColor: '#8BA0AB',
    borderBottomWidth: 1,
  },
  title: {
    fontWeight: 'bold',
    fontSize: 20,
  },
});

CRenderItem.propTypes = {
  item: Proptypes.shape({
    id: Proptypes.number.isRequired,
    title: Proptypes.string.isRequired,
    body: Proptypes.string.isRequired,
    userId: Proptypes.number.isRequired,
  }),
};
