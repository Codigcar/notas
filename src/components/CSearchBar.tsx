import React from 'react';
import {View, Platform} from 'react-native';
import {SearchBar} from 'react-native-elements';
import {SearchBarBaseProps} from 'react-native-elements/dist/searchbar/SearchBar';
import PropTypes from 'prop-types';

 export const CSearchBar = ({
  dataList,
  search,
  setSearch,
  setFilteredDataSource,
}: any) => {
  const SafeSearchBar = SearchBar as unknown as React.FC<SearchBarBaseProps>;

  const searchFilterFunction = (text: string) => {
    if (text) {
      const newData = dataList.filter(function (item: any) {
        const itemData = item.title;
        return itemData.indexOf(text) > -1;
      });
      setFilteredDataSource(newData);
      setSearch(text);
    } else {
      setFilteredDataSource(dataList);
      setSearch(text);
    }
  };

  return (
    <View>
      <SafeSearchBar
        containerStyle={styles.containerBarraBusqueda}
        inputContainerStyle={styles.estiloBarraBusqueda}
        onChangeText={(text: string) => searchFilterFunction(text)}
        onClear={() => searchFilterFunction('')}
        placeholder="Buscar por tÃ­tulo..."
        value={search}
        platform={'android'}
        clearIcon={true}
        testID={'search'}
      />
    </View>
  );
};

export const styles = {
  containerBarraBusqueda: {
    backgroundColor: '#fff',
    borderTopColor: '#FFF',
    borderBottomColor: '#FFF',
    paddingHorizontal: 0,
    marginRight: 20,
    ...Platform.select({
      ios: {
        paddingVertical: 0,
        borderRadius: 10,
      },
    }),
  },
  estiloBarraBusqueda: {
    backgroundColor: 'white',
    borderRadius: 10,
    marginVertical: 0,
    ...Platform.select({
      ios: {},
      android: {
        marginLeft: 15,
        marginRight: 15,
        shadowOpacity: 0.39,
        shadowRadius: 13.97,
        elevation: 11,
      },
      default: {
        shadowColor: 'rgba(0,0,0, .2)',
        shadowOffset: {height: 0, width: 0},
        shadowOpacity: 1,
        shadowRadius: 1,
      },
    }),
  },
};

CSearchBar.propTypes = {
  dataList: PropTypes.array.isRequired,
  search: PropTypes.string.isRequired,
  setSearch: PropTypes.func.isRequired,
  setFilteredDataSource: PropTypes.func.isRequired,
}
