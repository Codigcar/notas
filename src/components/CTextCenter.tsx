import React from 'react';
import {View, Text, StyleSheet} from 'react-native';
import Proptypes from 'prop-types';

type Props = {
  title: string;
};

export const CTextCenter = ({title}: Props) => {
  return (
    <View style={styles.center}>
      <View>
        <Text>{title}</Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  center: {
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
});

CTextCenter.propTypes = {
    title: Proptypes.string.isRequired
}