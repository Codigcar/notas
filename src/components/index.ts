

export * from './CRenderItem';
export * from './CRenderList';
export * from './CSearchBar';
export * from './CTextCenter';