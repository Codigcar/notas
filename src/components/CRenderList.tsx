import React from 'react';
import {FlatList, RefreshControl } from 'react-native';
import {CRenderItem} from './CRenderItem';
import PropTypes from 'prop-types';
import {CTextCenter} from './CTextCenter';

type Props = {
  data: any;
  pullRefresh: () => Promise<void>;
  isLoadingPullRefresh: boolean;
  loadMore?: (() => void) | null;
  timeEndReached?: number | null;
};

export const CRenderList = ({
  data,
  loadMore = null,
  timeEndReached = null,
  pullRefresh,
  isLoadingPullRefresh,
}: Props) => {
  return (
    <FlatList
      testID='flat-list'
      data={data}
      keyExtractor={item => item.id.toString()}
      renderItem={({item}) => <CRenderItem {...item} />}
      onEndReached={loadMore}
      onEndReachedThreshold={timeEndReached}
      refreshControl={
        <RefreshControl
          refreshing={isLoadingPullRefresh}
          onRefresh={pullRefresh}
          tintColor={'red'}
          colors={['red']}
          testID={'refresh-control'}
        />
      }
      contentContainerStyle={{flexGrow: 1}}
      ListEmptyComponent={() => {
        return <CTextCenter title="Publicaciones no encontradas" />;
      }}
    />
  );
};

CRenderList.propTypes = {
  data: PropTypes.array.isRequired,
  pullRefresh: PropTypes.func.isRequired,
  isLoadingPullRefresh: PropTypes.bool.isRequired,
  loadMore: PropTypes.func,
  timeEndReached: PropTypes.number,
};
