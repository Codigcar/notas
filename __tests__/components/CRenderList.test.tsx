import React from 'react';
import {CRenderList} from '../../src/components/CRenderList';

const {render, fireEvent, act} = require('@testing-library/react-native');
describe('testing en <CRenderList />', () => {
  const publicationsListMock = [
    {
      userId: 1,
      id: 1,
      title: 'title',
      body: 'body',
    },
    {
      userId: 1,
      id: 2,
      title: 'title',
      body: 'body',
    },
    {
      userId: 1,
      id: 3,
      title: 'title',
      body: 'body',
    },
  ];

  const pullRefreshMock = jest.fn();
  const loadMoreMock = jest.fn();

  const isLoadingPullRefreshMock = false;
  const timeEndReachedMock = 0.2;

  beforeEach(() => jest.clearAllMocks());

  it('debe hacer match con el snapshot', () => {
    const component = render(
      <CRenderList
        data={publicationsListMock}
        pullRefresh={pullRefreshMock}
        isLoadingPullRefresh={isLoadingPullRefreshMock}
        loadMore={loadMoreMock}
        timeEndReached={timeEndReachedMock}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('debe renderizar la lista de publicaciones', async () => {
    const {queryAllByRole} = render(
      <CRenderList
        data={publicationsListMock}
        pullRefresh={pullRefreshMock}
        isLoadingPullRefresh={isLoadingPullRefreshMock}
        loadMore={loadMoreMock}
        timeEndReached={timeEndReachedMock}
      />,
    );

    expect(queryAllByRole('menuitem').length).toBe(publicationsListMock.length);
  });

  it('debe mostrar un mensaje cuando cuando no encuentra resultado', () => {
    const {getByText} = render(
      <CRenderList
        data={[]}
        pullRefresh={pullRefreshMock}
        isLoadingPullRefresh={isLoadingPullRefreshMock}
        loadMore={loadMoreMock}
        timeEndReached={timeEndReachedMock}
      />,
    );

    const message = 'Publicaciones no encontradas';
    expect(getByText(message)).toBeTruthy();
  });

  it('debe ejecutar la funcion loadMoreMock al realizar un scroll para obtener el listado de publicaciones del siguiente usuario', async () => {
    const {getByTestId} = render(
      <CRenderList
        data={publicationsListMock}
        pullRefresh={pullRefreshMock}
        isLoadingPullRefresh={isLoadingPullRefreshMock}
        loadMore={loadMoreMock}
        timeEndReached={timeEndReachedMock}
      />,
    );

    const flatListComponent = getByTestId('flat-list');

    const eventData = {
      nativeEvent: {
        contentOffset: {
          y: 500,
        },
        contentSize: {
          height: 500,
          width: 100,
        },
        layoutMeasurement: {
          height: 100,
          width: 100,
        },
      },
    };

    fireEvent.scroll(flatListComponent, eventData);
    expect(loadMoreMock).toHaveBeenCalled();
  });

  it('debe ejecutar la funcion pullRefreshMock al realizar un dropdown', async () => {
    const {getByTestId} = render(
      <CRenderList
        data={publicationsListMock}
        pullRefresh={pullRefreshMock}
        isLoadingPullRefresh={isLoadingPullRefreshMock}
        loadMore={loadMoreMock}
        timeEndReached={timeEndReachedMock}
      />,
    );

    const flatListComponent = getByTestId('flat-list');
    expect(flatListComponent).toBeDefined();

    const {refreshControl} = flatListComponent.props;
    await act(async () => {
      refreshControl.props.onRefresh();
    });
    expect(pullRefreshMock).toHaveBeenCalled();
  });
});
