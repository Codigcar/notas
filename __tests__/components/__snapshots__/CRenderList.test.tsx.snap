// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing en <CRenderList /> debe hacer match con el snapshot 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  data={
    Array [
      Object {
        "body": "body",
        "id": 1,
        "title": "title",
        "userId": 1,
      },
      Object {
        "body": "body",
        "id": 2,
        "title": "title",
        "userId": 1,
      },
      Object {
        "body": "body",
        "id": 3,
        "title": "title",
        "userId": 1,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[MockFunction]}
  onEndReachedThreshold={0.2}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      colors={
        Array [
          "red",
        ]
      }
      onRefresh={[MockFunction]}
      refreshing={false}
      testID="refresh-control"
      tintColor="red"
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  testID="flat-list"
  viewabilityConfigCallbackPairs={Array []}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="menuitem"
        style={
          Object {
            "borderBottomColor": "#8BA0AB",
            "borderBottomWidth": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          title
        </Text>
        <Text>
          body
        </Text>
        <Text>
          User: 
          1
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="menuitem"
        style={
          Object {
            "borderBottomColor": "#8BA0AB",
            "borderBottomWidth": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          title
        </Text>
        <Text>
          body
        </Text>
        <Text>
          User: 
          1
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="menuitem"
        style={
          Object {
            "borderBottomColor": "#8BA0AB",
            "borderBottomWidth": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          title
        </Text>
        <Text>
          body
        </Text>
        <Text>
          User: 
          1
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
